prow/crier/reporters/github/reporter.go

```diff
@@ -68,6 +91,33 @@ func (c *Client) ShouldReport(pj *v1.ProwJob) bool {
 
 // Report will report via reportlib
 func (c *Client) Report(pj *v1.ProwJob) ([]*v1.ProwJob, error) {
+
+	// The github comment create/update/delete done for presubmits
+	// needs pr-level locking to avoid racing when reporting multiple
+	// jobs in parallel.
+	if pj.Spec.Type == v1.PresubmitJob {
+		key, err := lockKeyForPJ(pj)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get lockkey for job: %v", err)
+		}
+		lock := c.prLocks.getLock(key)
+		lock.Lock()
+		defer lock.Unlock()
+	}
+
 	// TODO(krzyzacy): ditch ReportTemplate, and we can drop reference to config.Getter
 	return []*v1.ProwJob{pj}, report.Report(c.gc, c.config().Plank.ReportTemplate, *pj, c.config().GitHubReporter.JobTypesToReport)
 }
+
+func lockKeyForPJ(pj *v1.ProwJob) (string, error) {
+	if pj.Spec.Type != v1.PresubmitJob {
+		return "", fmt.Errorf("can only get lock key for presubmit jobs, was %q", pj.Spec.Type)
+	}
+	if pj.Spec.Refs == nil {
+		return "", errors.New("pj.Spec.Refs is nil")
+	}
+	if n := len(pj.Spec.Refs.Pulls); n != 1 {
+		return "", fmt.Errorf("prowjob doesn't have one but %d pulls", n)
+	}
+	return fmt.Sprintf("%s/%s#%d", pj.Spec.Refs.Org, pj.Spec.Refs.Repo, pj.Spec.Refs.Pulls[0].Number), nil
```

Comment:
Is there a good reason for turning this into a string instead of having a simple triplet struct to serve as a map key?

Reply:
Yeah, I don't need to define a struct. What would be the advantage of doing that over the string, one `printf` computation less?

Reply:
I don't care about the performance of `Sprintf` (or even the key hashing in the map). I just like to use type system. Having an explicit simple struct to serve as a key IMO makes all these function signatures carry more information about their purpose. 

I don't insist.

Reply:
Added it.